### DISCUSSION TITLE
`pnpm list --recursive` should be sorted topologically

### DISCUSSION THREAD

**Discussion by mmkal** [pnpm/pnpm#15937] [Ideas]
### Contribution

- [x] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md)) (with some pointers in the right direction)

### Describe the user story

(Note: I'm on pnpm v8.5.1 - if this is fixed in another version pls let me know!)

When running `pnpm list --recursive`, I'd like to get the results back in topologically sorted order. For example, writing a script to publish packages, I want to make sure that I publish `lib-a` before `lib-b` which depends on it, and that before `app-c` which depends on `lib-a` and `lib-b`.

Easy repro, using trpc:

```
git clone https://github.com/trpc/trpc
cd trpc
pnpm install
pnpm list --recursive --filter './packages/*' --json --prod --only-projects
```

Note that the projects are listed in alphabetical order (so `@trpc/client` comes before `@trpc/server` even though the client has the server as a peer dependency). Run `pnpm recursive --filter './packages/*' exec pwd` to see the correct order.

### Describe the solution you'd like

I would have expected `pnpm list` to work like this already. The docs site for [`--recursive` / `--sort`](https://pnpm.io/cli/recursive#--no-sort) say that topological sorting is on by default. And other commands do indeed sort this way (e.g. `pnpm recursive exec pwd`). My current workaround is this:

```js
const list = await execa('pnpm', ['list', '--json', '--recursive', '--only-projects', '--prod', '--filter', './packages/*'])

ctx.packages = JSON.parse(list.stdout)

const pwdsCommand = await execa('pnpm', ['recursive', 'exec', 'pwd'])
const pwds = pwdsCommand.stdout.split('\n').map(s => s.trim()).filter(Boolean)

ctx.packages.sort((a, b) => pwds.indexOf(a.path) - pwds.indexOf(b.path))
```

So ideally this would just be "fixed". But if that's considered a breaking change, have `pnpm list` respect the `--sort` flag.

### Describe the drawbacks of your solution

If it's the new automatic behaviour: it's a breaking change, people may have scripts that _depend_ on `pnpm list` returning an alphabetically-sorted list.

If it requires `--sort`: it's confusing, since the docs say that `--sort` defaults to `true`.

### Describe alternatives you've considered

The workaround above. Using some pnpm APIs would probably work too but I'm aiming to have a simple shell wrapper right now.

**Comment by JAVA-LW** [comment]
I opened a PR for this. It orders recursive `pnpm list` results by workspace dependencies using the same graph sequencer as other recursive commands. The change covers JSON and text output, shared and dedicated lockfiles, and keeps `--no-sort` and `--reverse` available.

PR: https://github.com/pnpm/pnpm/pull/15855

---
Written by an agent (Codex, GPT-6).
